[WIP] Add Storage Transformers to Core Spec & Sharding v1.0 (#1)
scalableminds/zarr-specs: docs/storage_transformers/sharding/v1.0.rst

.. code-block::

    {
      storage_transformers: [
        {
          "storage_transformer": "https://purl.org/zarr/spec/storage_transformers/sharding/1.0",
          "configuration": {
            "format": "indexed",
            "chunks_per_shard": [
                2,
                2
            ]
          }
      ]
    }

``format``

    Specifies a `Binary shard format`_. In this version, the only binary format is the
    ``indexed`` format.

``chunks_per_shard``

    An array of integers providing the number of chunks that are combined in a shard
    for each dimension of the Zarr array, where each chunk may only start at a position
    that is divisble by ``chunks_per_shard`` per dimension, e.g. starting at the zero-origin.
    The length of the array must match the length of the array metadata ``shape`` entry.
    For example, a value ``[32, 2]`` indicates that 64 chunks are combined in one shard,
    32 along the first dimension, and for each of those 2 along the second dimension.
    Valid starting positions for a shard in the chunk-grid are therefore ``[0, 0]``,
    ``[32, 2]``, ``[32, 4]``, ``[64, 2]`` or ``[96, 18]``.

> Key & value transformation

Reviewer: This is basically the storage transformer implementation for sharding, right?
Maybe reflect that in this heading.

Reviewer: Key & value transformation with sharding storage transformer?

Author: Instead I added `Storage transformer implementation` as another heading and put the other two as subsections:

```
Storage transformer implementation
==================================

Key & value transformation
--------------------------

…

Binary shard format
-------------------

…
```

==========================

The storage transformer protocol defines the abstract interface to be the same
as the :ref:`abstract-store-interface`.

The Zarr store interface is defined in terms of `keys` and `values`,
where a `key` is a sequence of characters and a `value` is a sequence
of bytes. A key-value pair is called `entry` in the following part.

This sharding transformer only adapts entries where the key starts
with `data/root`, as they indicate data keys for array chunks. All other
entries are simply passed on.

Entries starting with ``data/root`` are grouped by their common shard, assuming
storage keys from a regular chunk grid which may use a customly configured
``chunk separator``:
For all entries that are part of the same shard the key is changed to the
shard-key and the values are combined in the `Binary shard format`_ described
below. The new shard-key is the chunk key divided by ``chunks_per_shard`` and

> floored per dimension. E.g. for ``chunks_per_shard=[32, 2]``, the chunk grid

Reviewer: Maybe start the example similar as above with: For example...

position ``[96, 18]`` (e.g. key "data/root/foo/baz/c96/18") is transformed to
the shard grid position ``[3, 9]`` and reassigned to the respective new key,
honoring the original chunk separator (e.g. "data/root/foo/baz/c3/9").
Chunk grid positions ``[96, 19]``, ``[97, 18]``, …, up to ``[127, 19]`` will
also have the same shard grid position ``[3, 9]``.

Binary shard format
===================

The only binary format is the ``indexed`` format, as specified by the ``format``
configuration key. Other binary formats might be added in future versions.

In the indexed binary format chunks are written successively in a shard, where
unused space between them is allowed, followed by an index referencing them.

> The index holds an `offset, length` pair of little-endian uint64 per chunk,

Reviewer: At which offset do I find the index?

Author: At the end of the file, I added an explanation.

The index always has the full shape of all possible chunks per shard,
even if they are outside of the array size.

The actual order of the chunk-content is not fixed and may be chosen by the implementation
as all possible write orders are valid according to this specification and therefore can
be read by any other implementation. When writing partial chunks into an existing shard no
specific order of the existing chunks may be expected. Some writing strategies might be

* **Fixed order**: Specify a fixed order (e.g. row-, column-major or Morton order).
  When replacing existing chunks larger or equal sized chunks may be replaced in-place,
  leaving unused space up to an upper limit  which might possibly be specified.
  Please note that for regular-sized uncompressed data all chunks have the same size and
  can therefore be replaced in-place.
* **Append-only**: Any chunk to write is appended to the existing shard, followed by an updated index.

Any configuration parameters for the write strategy must not be part of the metadata document,
in a shard I'd propose to use Morton order, but this can easily be changed and customized, since any order can be read.

> 

Reviewer: This sounds like an opinion, which should not be part of the spec, what do you think?

Author: Good point, that was just a left-over from a copy and paste, removed it and adapted the previous sentence

.. _sharding-storage-transformer-v1:

==========================================
Sharding storage transformer (version 1.0)
==========================================
-----------------------------
 Editor's draft 18 02 2022
-----------------------------

Specification URI:
    @@TODO
    http://purl.org/zarr/spec/storage_transformers/sharding/1.0
Issue tracking:
    `GitHub issues <https://github.com/zarr-developers/zarr-specs/labels/storage_transformers-sharding-v1.0>`_
Suggest an edit for this spec:
    `GitHub editor <https://github.com/zarr-developers/zarr-specs/blob/core-protocol-v3.0-dev/docs/storage_transformers/sharding/v1.0.rst>`_

Copyright 2022 `Zarr core development
team <https://github.com/orgs/zarr-developers/teams/core-devs>`_ (@@TODO
list institutions?). This work is licensed under a `Creative Commons
Attribution 3.0 Unported
License <https://creativecommons.org/licenses/by/3.0/>`_.

----

Abstract
========

This specification defines an implementation of the Zarr
storage transformer protocol for sharding.

Motivation
==========

> Sharding decouples the concept of chunks from storage keys, which become shards.

Reviewer 2: That is not entirely true. Shards hold 0 or more chunks, but not partial chunks.

A definition of sharding could be "co-locating multiple chunks within a file/storage object"

This is helpful when the requirements for those don't align:

> - Compressible units of chunks often need to be read and written in smaller
>   chunks, whereas

Reviewer 2: ```suggestion
- Chunk sizes need to be small for read efficiency requirements, e.g. for data streaming in browser-based visualization software, whereas
```

writing is not really facilitated through sharding.

- storage often is optimized for larger data per entry and fewer entries, e.g.
  as restricted by the file block size and maximum inode number for typical

>   file systems.

Reviewer 2: ```suggestion
- it becomes inefficient or impractical to store a large number of chunks in single files or objects due to the design constraints of the underlying storage, e.g. as restricted by the file block size and maximum inode number for typical file systems.
```

> This does not necessarily fit the access patterns of the data, so chunks might
> need to be smaller than one storage key. In those cases sharding decouples those
> entities. One shard corresponds to one storage key, but can contain multiple chunks:

Reviewer 2: ```suggestion
This does not necessarily fit the access patterns of the data, so chunks might
need to be smaller than the minimum size of one storage key. In those cases sharding decouples those
entities. One shard corresponds to one storage key, but can contain multiple chunks:
```

            "chunks_per_shard": [
                2,
                2
            ]
          }
      ]
    }

``format``

    Specifies a `Binary shard format`_. In this version, the only binary format is the
    ``indexed`` format.

``chunks_per_shard``

    An array of integers providing the number of chunks that are combined in a shard
    for each dimension of the Zarr array, where each chunk may only start at a position
    that is divisble by ``chunks_per_shard`` per dimension, e.g. starting at the zero-origin.
    The length of the array must match the length of the array metadata ``shape`` entry.
    For example, a value ``[32, 2]`` indicates that 64 chunks are combined in one shard,
    32 along the first dimension, and for each of those 2 along the second dimension.
    Valid starting positions for a shard in the chunk-grid are therefore ``[0, 0]``,
    ``[32, 2]``, ``[32, 4]``, ``[64, 2]`` or ``[96, 18]``.

Storage transformer implementation
==================================

Key & value transformation
--------------------------

The storage transformer protocol defines the abstract interface to be the same
as the :ref:`abstract-store-interface`.

> The Zarr store interface is defined in terms of `keys` and `values`,

Reviewer 2: ```suggestion
The Zarr store interface is defined as a mapping of `keys` and `values`,
```

where a `key` is a sequence of characters and a `value` is a sequence
of bytes. A key-value pair is called `entry` in the following part.

This sharding transformer only adapts entries where the key starts

> with `data/root`, as they indicate data keys for array chunks. All other

Reviewer 2: where does that come from? is that a constant or an example? can this be referenced to another part of the spec?

Author: That's part of the storage-key spec, a constant. I added a reference

Entries starting with ``data/root`` are grouped by their common shard, assuming
storage keys from a regular chunk grid which may use a customly configured
``chunk separator``:
For all entries that are part of the same shard the key is changed to the
shard-key and the values are combined in the `Binary shard format`_ described
below. The new shard-key is the chunk key divided by ``chunks_per_shard`` and
floored per dimension. For example for ``chunks_per_shard=[32, 2]``, the chunk grid
position ``[96, 18]`` (e.g. key "data/root/foo/baz/c96/18") is transformed to
the shard grid position ``[3, 9]`` and reassigned to the respective new key,
honoring the original chunk separator (e.g. "data/root/foo/baz/c3/9").
Chunk grid positions ``[96, 19]``, ``[97, 18]``, …, up to ``[127, 19]`` will
also have the same shard grid position ``[3, 9]``.

Binary shard format
-------------------

The only binary format is the ``indexed`` format, as specified by the ``format``
configuration key. Other binary formats might be added in future versions.

In the indexed binary format chunks are written successively in a shard, where
unused space between them is allowed, followed by an index referencing them.

> The index is placed at the end of the file and has a length of 16 bytes per chunk

Reviewer 2: ```suggestion
The index is placed at the end of the file and has a length of 16 bytes multiplied by the number of chunks
```

The index always has the full shape of all possible chunks per shard,
even if they are outside of the array size.

The actual order of the chunk-content is not fixed and may be chosen by the implementation
as all possible write orders are valid according to this specification and therefore can
be read by any other implementation. When writing partial chunks into an existing shard no
specific order of the existing chunks may be expected. Some writing strategies might be

* **Fixed order**: Specify a fixed order (e.g. row-, column-major or Morton order).
  When replacing existing chunks larger or equal sized chunks may be replaced in-place,
  leaving unused space up to an upper limit  which might possibly be specified.
  Please note that for regular-sized uncompressed data all chunks have the same size and
  can therefore be replaced in-place.
* **Append-only**: Any chunk to write is appended to the existing shard,
  followed by an updated index.

Any configuration parameters for the write strategy must not be part of the metadata document,
they need to be configured at runtime, as this is implementation specific.

> 

Reviewer 2: I think what is missing is a read and write protocol, i.e. how does reading/writing work in an abstract algorithm.

    `GitHub editor <https://github.com/zarr-developers/zarr-specs/blob/core-protocol-v3.0-dev/docs/storage_transformers/sharding/v1.0.rst>`_

Copyright 2022 `Zarr core development
team <https://github.com/orgs/zarr-developers/teams/core-devs>`_ (@@TODO
list institutions?). This work is licensed under a `Creative Commons
Attribution 3.0 Unported
License <https://creativecommons.org/licenses/by/3.0/>`_.

----

Abstract
========

This specification defines an implementation of the Zarr
storage transformer protocol for sharding.

Sharding co-locates multiple chunks within a storage object, bundling them in shards.

Motivation
==========

In many cases it becomes inefficient or impractical to store a large number of chunks in
single files or objects due to the design constraints of the underlying storage,
for example as restricted by the file block size and maximum inode number for typical file systems.

Increasing the chunk size works only up to a certain point, as chunk sizes need to be small for
read efficiency requirements, for example to stream data in browser-based visualization software.

Therefore, chunks may need to be smaller than the minimum size of one storage key.

> In those cases it is required to store objects at a more coarse granularity than reading chunks.

Reviewer 2: ```suggestion
In those cases it is efficient to store objects at a more coarse granularity than reading chunks.
```

The term `underlying store` references either the next storage transformer
in the stack or the actual store if this transformer is the last one in the
stack. Any operations with keys not starting with ``data/root`` are simply
relayed to the underlying store and not described explicitly.

* ``get_partial_values(key_ranges) -> values``:
  For each referenced key, request the indices from the underlying store using
  ``get_partial_values``. For each `key`, `range` pair in in `key_ranges`,
  check if the chunk exists by checking if the index offset and length
  are both ``2^64 - 1``. For existing keys, request the actual chunks by
  their ranges as read from the index using ``get_partial_values``.
  This operation should be implemented using two ``get_partial_values``
  operations on the underlying store, one for retrieving the indices and
  one for retrieving existing chunks.

* ``set_partial_values(key_start_values)`` :
  For each referenced key, check if all available chunks in a shard are
  referenced. In this case a shard can be constructed according to the
  `Binary shard format`_ directly.
  For all other keys, request the indices from the underlying store using

>   ``get_partial_values``. All chunks that are not updated completely and

Reviewer: how does the `get_partial_values` function return the indices?

Author: The index format is described above.

Reviewer: So you use (store or next transformer).get_partial_values to read the index to perform the operation of the transformer.get_partial_values function?
To me, this is confusing as they are called the same but are defined on the transformer and on the store, right?

Reviewer: But I guess it makes sense :)

Author: > So you use (store or next transformer).get_partial_values to read the index to perform the operation of the transformer.get_partial_values function?

Exactly :+1: 

> To me, this is confusing as they are called the same but are defined on the transformer and on the store, right?

Yes, a bit, but that's also part of the beauty, as nobody cares if the "next" store is actually the store or just the next transformer. I tried to always use the term `underlying store` refering to the next transformer or store, as described in the paragraph above in the spec:
> The term `underlying store` references either the next storage transformer
in the stack or the actual store if this transformer is the last one in the

However, it's still confusing, but I have no idea how to improve it. Do you have a suggestion?
